pkg/middleware/middleware.go

```diff
@@ -43,60 +47,200 @@ func AuthenticationMiddleware() gin.HandlerFunc {
+			c.Set("role", *user.Userlevel)
+		} else if iss == os.Getenv("OIDC_ISSUER") {
+			if auth.Jwks == nil || os.Getenv("OIDC_USERNAME_KEY") == "" {
+				log.Print("\033[31mError: OIDC environment variables not configured properly\033[0m")
+				er := models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "Something went wrong",
+					Error:     "Something went wrong",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusInternalServerError, er)
+				c.Abort()
+				return
+			}
+
+			keyset, err := auth.Jwks.Lookup(context.Background(), os.Getenv("JWKS_URI"))
+			if err != nil {
+				log.Print("\033[31mError: Failed to create a jwk.Cache from the oidc provider's URL\033[0m")
+				er := models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "Something went wrong",
+					Error:     "Something went wrong",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusInternalServerError, er)
+				c.Abort()
+				return
+			}
+
+			parsedToken, err := jwt.Parse([]byte(tokenString), jwt.WithValidate(true), jwt.WithVerify(false))
+			if err != nil {
+				fmt.Println(err.Error())
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "Please check your credentials and try again",
+					Error:     "Please check your credentials and try again",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusUnauthorized, er)
+				c.Abort()
+				return
+			}
+
+			keyOptions := jws.WithKeySet(keyset, jws.WithInferAlgorithmFromKey(true))
+			kid := ""
+			if err := parsedToken.Get("kid", &kid); err != nil {
+				if key, ok := keyset.LookupKeyID(kid); ok {
+					if os.Getenv("OIDC_SIGNING_ALG") != "" {
+						if alg, ok := jwa.LookupSignatureAlgorithm(os.Getenv("OIDC_SIGNING_ALG")); ok {
+							key.Set("alg", alg)
+							keyOptions = jws.WithKeySet(keyset)
+						}
+					} else if _, ok := key.Algorithm(); ok {
+						keyOptions = jws.WithKeySet(keyset)
+					}
+				}
+			}
+
+			if _, err = jws.Verify([]byte(tokenString), keyOptions); err != nil {
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "Please check your credentials and try again",
+					Error:     "Please check your credentials and try again",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusUnauthorized, er)
+				log.Printf("\033[31mError: Token verification \033[0m")
+				return
+			}
+
+			var username string
+			if err = parsedToken.Get(os.Getenv("OIDC_USERNAME_KEY"), &username); err != nil {
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "Please check your credentials and try again",
+					Error:     "Please check your credentials and try again",
```

Comment:
Having the same content in error and message defeats its purpose, unless we want to redact confidential data. There are multiple instances which the same content.